recipes/generate.py

```diff
@@ -81,9 +81,40 @@ def _setup_model(
 
         return model
 
+    def convert_prompt_to_tokens(
+        self,
+        prompt: Union[DictConfig, str],
+        chat_format: ChatFormat,
+        instruct_template: InstructTemplate,
+    ) -> List[Message]:
+        # Should only be chat-style prompt or instruct-style prompt
+        if chat_format and instruct_template:
+            raise ValueError(
+                "Cannot pass both chat format and instruct template for generation"
+            )
+
+        # A single string can be formatted with instruct_template
+        if isinstance(prompt, str):
+            if instruct_template:
+                prompt = template.format(prompt)
+            return self._tokenizer.encode(prompt, add_bos=True, add_eos=False)
+
+        # Otherwise pass a DictConfig mapping role -> message, format with ChatFormat
+        # dict.items() will respect order for Python >= 3.7
+        else:
+            messages = [Message(role=k, content=v) for k, v in prompt.items()]
+            messages += [
+                Message(role="assistant", content="")
+            ]  # TODO: pass via config?
+            if chat_format:
+                messages = chat_format.format(messages)
```

Comment:
doing this after appending the assistant message is slightly problematic because the chat format may add tags at the end of the assistant message (like ChatML, [for example](https://github.com/pytorch/torchtune/blob/3b1bcf9f084ae7a03fd20251402e0292fa9a9495/torchtune/data/_chat_formats.py#L190)), but ideally we just want the model to output tokens starting with where the assistant message should start. I don't have a good solution to this tbh...

Reply:
Yeah that's kinda annoying. Where are we actually even using `ChatMLFormat` though?

Reply:
nowhere really, I added it because at the time it was a widely used prompt template used by OpenAI. not sure if that is still true today.